### ISSUE TITLE
Improve Target Build functionality

### ISSUE THREAD

**Issue by gene-pavlovsky** [fdorg/flashdevelop#1181]
@elsassph Thanks, since it's a drop-down, I didn't figure I could just type something in there. This raises a bunch more questions, perhaps we should have this conversation elsewhere?
The way Target Build dropdown currently works, it seems that every project remembers the last value of target build. 
If I type some target name, build, then type another, build, I can see both of the targets in the dropdown. 
It doesn't look possible to remove items from there, or to persist the entire list - when I restart FD or close/open project - only the last target is shown, and dropdown is empty. 
Additionally, looks like the last target is stored somewhere in FD local app data, not the project properties files. 

1. I think available target builds should be editable in project properties, and stored in the project xml file. Are there plans to implement 

2. `fdbuild` is called with -target $(TargetBuild) - however doesn't look like that is doing anything further. Shouldn't it at least define a CONFIG::$(TargetBuild) compiler constant, or something like that?

3. If I delete all the text in the target build dropdown, `fdbuild` is called with `-target ""` instead of no target. How to reset target to null?

4. Is it possible to call a custom command instead of `fdbuild`? I saw a `custom build` project properties option, but can't figure out where to put my custom build command (unless I'm supposed to put it in pre-build or post-build). Actually it would be preferrable if I could have default `fdbuild` behavior with one target and a custom command (e.g. script) with another target, are there any plans like this? I know you're working on #548 so all this will probably be replaced or significantly changed anyway.

**Comment by elsassph** [comment]
1. yes we have a few ideas like that
2. currently the argument is for you to use, but we could certainly imagine adding a compiler constant - we didn't think about that
3. minor bug, null or empty have the same effect
4. if you choose Custom Build nothing should be built automatically and you can instead define the command to execute in the Build tab > Pre/post-build

**Comment by gene-pavlovsky** [comment]
All right, thanks for clearing it up.
2. Is there any way to affect how fdbuild operates? E.g. besides the options offered in project properties, can I insert some script in between fdbuild and calling of the compiler? Or is there any way I can run a script to modify the auto-generated build config xml (obj/ProjectNameConfig.xml)? Is that xml generated before or after the pre-build command is run?
4. So how about if Custom Build is selected, `Pre-Build Command Line` label should instead show `Build Command Line`?

**Comment by elsassph** [comment]
2. You can use the argument in a pre-build script, but this happens before the Flex config is generated. 
4. You're right, it would be friendlier. 

Everything we discussed should be added to the wiki also.

**Comment by gene-pavlovsky** [comment]
Ok, do you think this would work? I didn't test this, but if everything works the way I suppose it does, this should set `CONFIG::ios` or `CONFIG::android` compiler constant if target is `ios` or `android`, respectively, enabling to use conditional compilation rather than run-time checking.

1. Set compiler option `Load configuration file` to `obj/ProjectName-target.xml` (for example).
2. Set pre-build command line: `scripts/setup-target.sh $(TargetBuild)`
3.
`scripts/setup-target.sh` (bash script in this example):
```bash
#!/bin/sh
ios_config=build/ios-config.xml
android_config=build/android-config.xml

target_config=obj/ProjectName-target.xml

case $1 in
	ios)
		cp -f "$ios_config" "$target_config"
	;;
	android)
		cp -f "$android_config" "$target_config"
	;;
	'')
		rm -f "$target_config"
	;;
	*)
		echo "Unknown target \"$1\"" >&2
		exit 1
	;;
esac
```

In this minimal example the script just copies a target-specific compiler configuration file, which can contain all the necessary options for a particular target.

`build/ios-config.xml`:
```
<?xml version="1.0" encoding="utf-8"?>
<flex-config>
  <compiler>
    <define append="true"><name>CONFIG::android</name><value>false</value></define>
    <define append="true"><name>CONFIG::ios</name><value>true</value></define>
  </compiler>
</flex-config>
```

`build/android-config.xml`:
```
<?xml version="1.0" encoding="utf-8"?>
<flex-config>
  <compiler>
    <define append="true"><name>CONFIG::android</name><value>true</value></define>
    <define append="true"><name>CONFIG::ios</name><value>false</value></define>
  </compiler>
</flex-config>
```

**Comment by elsassph** [comment]
Yes this should definitely work and is basically the use case for the target build option.

**Comment by gene-pavlovsky** [comment]
I like the idea of putting use cases / examples in the wiki. I bet there are many nice FD features many people would use if they only knew about them / how to use them.
